### PR TITLE
Bug fix for muon results table and coadd runs

### LINKED ISSUE #24237: MuonAnalysis results table not properly retrieving some values
The MuonAnalysis GUI seems to currently not be retrieving some of the required values to show in the results table. There are two variants of this: 
 * One is that for a co-added set of runs the when you create a results table selecting run-number and sample_temp as the logs of interest these two columns contains zeros rather than the relevant data.
 * The other is that when you select to display the group for a sequential or simultaneous fit the group column displays a 0 rather than the relevant group.

### PR DESCRIPTION
**Pull request by AnthonyLim23**
**Description of work.**
The results table of muon analysis was returning zeros in some of the cells as it was casting strings into numbers. 
**Report to:** Matthew and Keith. 

**To test:**
Open muon analysis
load a run
go to the settings tab
make sure multiple fitting is on
go to the data analysis tab
add a fitting function
in the runs box add a few more (e.g. 60212-6)
Make sure co-add is selected
Change the selected groups/pairs to groups
Do a fit
Go to the results table tab
Select multiple (middle of page)
Pick some sample logs (e.g. groups, sample_temp)
Click create table
Check it has sensible values

Click simultaneous (middle of page)
Click create table
check it has sensible values

Fixes #24237. 

In release notes
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp [hunk 1]
```diff
@@ -129,9 +139,13 @@ ITableWorkspace_sptr MuonAnalysisResultTableCreator::createTable() const {
   for (const auto &log : m_logs) {
 
     auto val = valMap[log];
-    // multiple files use strings due to x-y format
-    if (val.canConvert<double>() && !log.endsWith(" (text)")) {
+    auto dashIndex = val.toString().indexOf("-");
 
+    // multiple files use strings due to x-y format
+    if (dashIndex != 0 && dashIndex != -1) {
+      addColumnToTable(table, "str", log.toStdString(), PLOT_TYPE_X);
```
Comment (Mantid-Matthew) [inline qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp]: This is the same expression as the else option below, can the two be combined in some way? As a more general point is there any reason we can't get the values of everything as strings and add them to the table this way using `addColumnToTable` for all cases? This point also applies to some of the cases below.

File: qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp [hunk 2]
```diff
@@ -24,7 +26,15 @@ using Mantid::API::WorkspaceFactory;
 using Mantid::API::WorkspaceGroup;
 
 namespace {
-
+/**
+ * Checks if a string is a numeric value
+ * @param string:: string to test
+ * @returns :: bool if it is a string
+ */
+bool isNumber(const std::string &string) {
```
Comment (Mantid-Matthew) [inline qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp]: Could we add some unit tests to this to help document, what the regex does and does not reject?

File: qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp [hunk 3]
```diff
@@ -541,13 +556,15 @@ void MuonAnalysisResultTableCreator::addColumnToResultsTable(
   for (const auto &wsName : paramsByLabel[labelName].keys()) {
     const auto &logValues = m_logValues->value(wsName);
     const auto &val = logValues[log];
-
+    auto a = log.toStdString();
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp]: ah no, I missed them when cleaning up

File: qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp (hunk 1 above)
Comment (AnthonyLim23) [reply qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp]: Unfortunately they want numeric columns when possible, so they can be easily plotted. We could remove the else by inverting the following if statements and making them part of this one.

File: qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp (hunk 2 above)
Comment (AnthonyLim23) [reply qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp]: yes, I can quickly put something together

File: qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp (hunk 1 above)
Comment (AnthonyLim23) [reply qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp]: Looking at it I think trying to condense will probably lead to more bugs

**Review by Matthew-Andrew — COMMENTED** [review]
This now produces the correct values in the table for me. There seems to be some build warnings on windows to deal with however.

File: qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp (hunk 2 above)
Comment (martyngigg) [reply qt/scientific_interfaces/Muon/MuonAnalysisResultTableCreator.cpp]: The Windows compiler is warning on the fact that we've included a `.cpp` in a test header and we're trying to then link to something in an anonymous namespace - which you're not supposed to do as they are supposed to be entirely internal. Anonymous methods shouldn't be tested directly as they should be considered implementation details. If something is big/complicated enough to require a test it should be it's own function/class with accompanying tests.

I would also suggest not using a `regex` for this sort of check as this type of functionality already exists in things like [`QString::toDouble`](http://doc.qt.io/archives/qt-4.8/qstring.html#toDouble) et al or `boost::lexical_cast`. As you already have a `QString` from `toString` I would suggest using the `QString` methods as they will be proven methods of getting edge cases in this type of thing correct.

**Review by Matthew-Andrew — APPROVED** [review]
This looks good to go now :+1:

**Pull request merged**